bpo-34002: Minor efficiency and clarity improvements in email package.

selik opened the pull request:
Creating a copy of a list, then repeatedly popping the first element could be much slower than necessary if the list is long. Further explanations in the commit messages

Automerge-Triggered-By: @maxking

selik committed:
Check intersection of two sets explicitly

Comparing ``len(a) > ``len(a - b)`` is essentially looking for an
intersection between the two sets. If set ``b`` does not intersect ``a``
then ``len(a - b)`` will be equal to ``len(a)``. This logic is more
clearly expressed as ``a & b``.

selik committed:
Change while/pop to a for-loop

Copying the list, then repeatedly popping the first element was
unnecessarily slow. I also cleaned up a couple other inefficiencies.
There's no need to unpack a tuple, then re-pack and append it. The list
can be created with the first element instead of empty. Secondly, the
``endswith`` method returns a bool, so there's no need for an if-
statement to set ``encoding`` to True or False.

serhiy-storchaka commented on Lib/email/headerregistry.py:
```diff
@@ -141,10 +141,8 @@ def __str__(self):
-            nameset = set(disp)
-            if len(nameset) > len(nameset-parser.SPECIALS):
-                disp = parser.quote_string(disp)
+        if disp is not None and set(disp) & parser.SPECIALS:
```
This can be written as
```python
if disp is not None and not parser.SPECIALS.isdisjoint(disp):
```

serhiy-storchaka replied:
There is also a similar code at line 90.

selik replied:
Updated the code to use ``set.intersection``, which I think reads better than "not isdisjoint". I spotted a 3rd similar comparison and changed that one as well.

I noted in the commit message that this avoids an unnecessary set construction.

serhiy-storchaka replied:
`set.intersection()` creates a new set, while `set.isdisjoint()` returns a boolean and stops fast if a common item is found.

serhiy-storchaka commented on Lib/email/utils.py:
```diff
@@ -259,21 +259,13 @@ def decode_params(params):
-    # Copy params so we don't mess with the original
-    params = params[:]
-    new_params = []
+    new_params = [params[0]]
```
What if `params[0]` is not a 2-tuple?

selik replied:
That's possible for the first params element, though the others are implicitly checked by the unpacking for-loop.

Looking at the usage of `decode_params`, it's only use is in `email.message._get_params_preserve` which in turn has 3 call sites. All 3 calls unpack the elements as `(k, v)` pairs. The bad data would cause an immediate error, given how the method is used. The error would be the same as today, though the traceback would not be as deep.

If that's not satisfactory, I'll revert to:

    key, value = params[0]
    new_params = [(key, value)]

serhiy-storchaka reviewed (commented):
Please open an issue on the tracker for discussing this PR.

selik committed:
Use set.intersection to check for intersections

``a.intersection(b)`` method is more clear of purpose than ``not
a.isdisjoint(b)`` and avoids an unnecessary set construction that ``a &
set(b)`` performs.

rhettinger commented on Lib/email/headerregistry.py:
```diff
@@ -69,11 +69,9 @@ def addr_spec(self):
-        nameset = set(self.username)
-        if len(nameset) > len(nameset-parser.DOT_ATOM_ENDS):
+        lp = self.username
+        if parser.DOT_ATOM_ENDS.intersection(lp):
```
Consider using `set.isdisjoint()`.

selik replied:
@serhiy-storchaka also suggested "not isdisjoint". Is that because it explicitly returns a `bool`?

selik replied:
GitHub hid @serhiy-storchaka's explanation after I pushed a change. Makes sense to stop early when finding a match via disjoint. I'll make the change ...

serhiy-storchaka commented on Lib/email/headerregistry.py:
```diff
@@ -69,11 +69,9 @@ def addr_spec(self):
-        if len(nameset) > len(nameset-parser.DOT_ATOM_ENDS):
+        lp = self.username
+        if parser.DOT_ATOM_ENDS.intersection(lp):
             lp = parser.quote_string(self.username)
```
You could use `lp` instead of `self.username`.

selik replied:
Fixed.

serhiy-storchaka commented on Lib/email/utils.py:
```diff
@@ -259,21 +259,13 @@ def decode_params(params):
-            encoded = True
-        else:
-            encoded = False
+    for name, value in params[1:]:
```
`params[1:]` still creates a copy of the list.

selik replied:
True. The real benefit is avoiding repeated `.pop(0)`.

serhiy-storchaka replied:
But the code can be made more optimal if avoid a copy.

terryjreedy replied:
It looks to me that the final value of 'encoded' only depends on the final value of name.  If so, the loop is equivalent to
```
if params:
    encoded = params[-1][0].endswith('*')
```
Otherwise, one can skip-iterate without copying with
```
it = iter(params)
try:
    next(it)
except StopIteration:
    pass
for name, value in it:
   ...
```

selik replied:
Each iteration of the loop uses the `encoded` value. The `iter` and `next`ing would be a better technique.

```
it = iter(params)
new_params = [next(it)]
for name, value in it:
    ...
```

selik committed:
Use not isdisjoint instead of intersection

While it reads slightly worse, the isdisjoint method will stop when it
finds a counterexample and returns a bool, rather than looping over the
entire iterable and constructing a new set.

maxking committed:
Add NEWS item

maxking reviewed (approved):
LGTM
